server/src/main/java/org/candlepin/controller/CandlepinOwnerProductShareManager.java

```diff
@@ -0,0 +1,200 @@
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.controller;
+
+import org.candlepin.model.Owner;
+import org.candlepin.model.OwnerProductCurator;
+import org.candlepin.model.OwnerProductShare;
+import org.candlepin.model.OwnerProductShareCurator;
+import org.candlepin.model.Product;
+
+import com.google.inject.Inject;
+
+import org.apache.commons.collections.CollectionUtils;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Manages Shared product operations.
+ */
+public class CandlepinOwnerProductShareManager implements OwnerProductShareManager {
+
+    private OwnerProductCurator ownerProductCurator;
+    private OwnerProductShareCurator shareCurator;
+
+    @Inject
+    public CandlepinOwnerProductShareManager(OwnerProductCurator ownerProductCurator,
+        OwnerProductShareCurator shareCurator) {
+        this.ownerProductCurator = ownerProductCurator;
+        this.shareCurator = shareCurator;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Product resolveProductById(Owner recipientOwner, String productId, boolean includeShares) {
+        List<Product> products = resolveProductsByIds(recipientOwner,
+            Collections.singleton(productId), includeShares);
+        if (CollectionUtils.isNotEmpty(products)) {
+            return products.get(0);
+        }
+        return null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<Product> resolveProductsByIds(Owner recipientOwner, Collection<String> productIds,
+        boolean includeShares) {
+        Map<String, Product> result = new HashMap<String, Product>();
+        List<Product> existingProducts;
+        Set<String> missingProducts = new HashSet<String>();
+
+        if (productIds != null && !productIds.isEmpty()) {
+            // add all requested now so we can remove the found ones later.
+            missingProducts.addAll(productIds);
+            existingProducts = ownerProductCurator.getProductsByIds(recipientOwner, productIds).list();
+        }
+        else {
+            existingProducts = ownerProductCurator.getProductsByOwner(recipientOwner).list();
+        }
+
+        for (Product product : existingProducts) {
+            result.put(product.getId(), product);
+            missingProducts.remove(product.getId());
```

Comment:
If you wanted to be fancy you could probably do this all in one go with something like

```
missingProducts = missingProducts.removeAll(result.getKeys());
```

That's pseudo-code, but I imagine there's some method to remove a bunch of items at once from a set and I think `getKeys` there would return a set.